### ISSUE TITLE
HTTP.request sidesteps a bit too much ?

### ISSUE THREAD

**Issue by fredrikekre** [JuliaWeb/HTTP.jl#463]
This new method: https://github.com/JuliaWeb/HTTP.jl/blob/7bcb6f554f116eb075a3286d641be8fbb2b400bf/src/HTTP.jl#L316 sidesteps a bit too much. I assume it is meant to call the method just above: https://github.com/JuliaWeb/HTTP.jl/blob/7bcb6f554f116eb075a3286d641be8fbb2b400bf/src/HTTP.jl#L306-L314 in order to correctly deal with `headers` kwargs etc. But that does not happen, because we dispatch to https://github.com/JuliaWeb/HTTP.jl/blob/7bcb6f554f116eb075a3286d641be8fbb2b400bf/src/RedirectRequest.jl#L18 which completely ignores `headers` (it is included in `kwargs...`).

Here is an example, where `headers` is ignored:
```
julia> using HTTP

julia> uri = HTTP.URI("https://api.github.com");

julia> headers=Dict("User-Agent" => "HTTP.jl");

julia> HTTP.request("GET", uri; headers=headers)
ERROR: HTTP.ExceptionRequest.StatusError(403, "GET", "/", HTTP.Messages.Response:
"""
HTTP/1.0 403 Forbidden
Cache-Control: no-cache
Connection: close
Content-Type: text/html

Request forbidden by administrative rules. Please make sure your request has a User-Agent header (http://developer.github.com/v3/#user-agent-required). Check https://developer.github.com for other possible causes.
""")
```

**Comment by oxinabox** [comment]
I cannot assign @mattBrzezinski to this issue because he isn't a collaborator.
But he says he is going to work on this now.

**Comment by mattBrzezinski** [comment]
I spent a bit of time looking into a solution for this, and using your test case from above this does not seem to be an issue with `v0.8.6` and the introducing for a secondary entry point.

```julia
(v1.1) pkg> add HTTP@0.8.1
julia> import Pkg; Pkg.installed()
Dict{String,Union{Nothing, VersionNumber}} with 22 entries:
  ...
  "HTTP"           => v"0.8.1"
  ...
julia> using HTTP
julia> uri = HTTP.URI("https://api.github.com");
julia>  headers=Dict("User-Agent" => "HTTP.jl");
julia> HTTP.request("GET", uri; headers=headers)
ERROR: HTTP.ExceptionRequest.StatusError(403, "GET", "/", HTTP.Messages.Response:
"""
HTTP/1.0 403 Forbidden
Cache-Control: no-cache
Connection: close
Content-Type: text/html

Request forbidden by administrative rules. Please make sure your request has a User-Agent header (http://developer.github.com/v3/#user-agent-required). Check https://developer.github.com for other possible causes.
""")
```

It looks like this has been around for a while now, I'll spend some more time looking into this later this week.

**Comment by iamed2** [comment]
The change was made [here](https://github.com/JuliaWeb/HTTP.jl/commit/a9c7866222c6c156635f4d70073983f2653c7a3c#diff-ecb509fe928282cfc7a3c93198c61e32R359) to add a stack argument to `HTTP.request`. However, this basically ensures that method will never be called, since a stack is just a `Layer`, so it will dispatch to the appropriate layer.

**Comment by quinnj** [comment]
What's the status here? @fredrikekre, is your PR ready to review? (it's currently marked as a draft). @mattBrzezinski, did you ever discover anything when you said you'd look into this?

**Comment by iamed2** [comment]
https://github.com/JuliaWeb/HTTP.jl/pull/470 resolved this particular issue, and it looks like https://github.com/JuliaWeb/HTTP.jl/pull/469 was noted as a breaking change

**Comment by fredrikekre** [comment]
Yes, I think (something like) #469 would be good for the next breaking release, it is just too difficult to separata public API from internal API when everything is methods of the same function.

**Comment by quinnj** [comment]
Resolved in #789
